narwhal/network/src/anemo_ext.rs

```diff
@@ -44,29 +47,42 @@ impl WaitingPeer {
         }
 
         // If we're not connected we'll need to check to see if the Peer is a KnownPeer
+        let timeout = request.timeout();
+        let sleep: OptionFuture<_> = timeout.map(tokio::time::sleep).into();
+        tokio::pin!(sleep);
         loop {
             if self.network.known_peers().get(&self.peer_id).is_none() {
                 return Err(format!("peer {} is not a known peer", self.peer_id).into());
             }
 
-            match subscriber.recv().await {
-                Ok(PeerEvent::NewPeer(peer_id)) if peer_id == self.peer_id => {
-                    // We're now connected with the peer, lets try to make a network request
-                    if let Some(mut peer) = self.network.peer(self.peer_id) {
-                        return peer.rpc(request).await.map_err(Into::into);
+            tokio::select! {
+                recv = subscriber.recv() => match recv {
+                    Ok(PeerEvent::NewPeer(peer_id)) if peer_id == self.peer_id => {
+                        // We're now connected with the peer, lets try to make a network request
+                        if let Some(mut peer) = self.network.peer(self.peer_id) {
+                            if let Some(duration) = timeout {
+                                // Reduce timeout to account for time already spent waiting
+                                // for the peer.
+                                request.set_timeout(duration.saturating_sub(Instant::now().duration_since(start)));
+                            }
+                            return peer.rpc(request).await.map_err(Into::into);
+                        }
                     }
-                }
-                Err(RecvError::Closed) => return Err("network is closed".into()),
-                Err(RecvError::Lagged(_)) => {
-                    subscriber = subscriber.resubscribe();
+                    Err(RecvError::Closed) => return Err("network is closed".into()),
+                    Err(RecvError::Lagged(_)) => {
+                        subscriber = subscriber.resubscribe();
 
-                    // We lagged behind so we may have missed the connection event
-                    if let Some(mut peer) = self.network.peer(self.peer_id) {
-                        return peer.rpc(request).await.map_err(Into::into);
+                        // We lagged behind so we may have missed the connection event
+                        if let Some(mut peer) = self.network.peer(self.peer_id) {
+                            return peer.rpc(request).await.map_err(Into::into);
+                        }
                     }
-                }
-                // Just do another iteration
-                _ => {}
+                    // Just do another iteration
+                    _ => {}
+                },
+                Some(_) = &mut sleep => {
```

Comment:
I think this is still going to be a busy loop if sleep is None:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f5daf418a83056e0a659fdfe75e3a291
You need a precondition here:
```
                Some(_) = &mut sleep if  timeout .is_some() => {
```

Reply:
That playground link seems to work fine

Reply:
I added some stuff to your playground link whose output seems to suggest to me that this does what we want and is not busy looping?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=65332d1ce5003e9bd7f7bc59dca42c95

Reply:
Ok, I see, I think you are right